java/memory/src/main/java/org/apache/arrow/memory/util/hash/DirectHasher.java

```diff
@@ -0,0 +1,82 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.arrow.memory.util.hash;
+
+/**
+ * Calculate hash code by directly returning the integers.
+ * This is the default and the fastest way to get the hash code.
+ * <p>
+ *   Objects of class are stateless, so it can be shared between threads.
+ * </p>
+ */
+public class DirectHasher extends ArrowBufHasher {
+
+  public static DirectHasher INSTANCE = new DirectHasher();
+
+  private static final int DEFAULT_SEED = 0;
+
+  private DirectHasher() {
+
+  }
+
+  @Override
+  public int combineHashCode(int currentHashCode, int newHashCode) {
+    return currentHashCode * 37 + newHashCode;
+  }
+
+  @Override
+  public int getByteHashCode(byte byteValue) {
+    return (int) byteValue;
+  }
+
+  @Override
+  public int getIntHashCode(int intValue) {
+    return intValue;
+  }
+
+  @Override
+  public int getLongHashCode(long longValue) {
+    return Long.hashCode(longValue);
+  }
+
+  @Override
+  public int finalizeHashCode(int hashCode) {
```

Comment:
@emkornfield Since we hashCode & equals API already checked in, could we take a look at this PR? Something like finalizeHashCode is useful in actual applications.

Reply:
@tianchen92 I'm not sure I understand the question exactly?  What were you thinking about doing with ti?

Reply:
I think algorithm like Murmur hashing will significantly reduce hash collision.

Reply:
It could, this code doesn't seem to implement full murmur hash though, I don't know how much just the finalization steps helps

Reply:
You are right, this is not full murmur hash. However, it is simple and effective:
1. the hash code on int/byte/long are poor (they are not spread evenly in the universe), but fast enough
2. the hash code finalized by murmur hash makes the result evenly distributed in the universe. This is costly, as there are a few integer multiplications. However, it is carried out only once.

Without the finalization step, the quality of the hash code can be poor. For example, suppose we have a integer array with small positive numbers (this is common in practice, and many cases can be converted to an equivalent one). Without the finalization step, the hash code would also be a small integer (though larger), so it is not evenly distributed in the universe, and may cause problems (e.g. in an open addressing hash table).